src/utils/daos_hdlr.c

```diff
@@ -2948,6 +2963,465 @@ fs_copy_hdlr(struct cmd_args_s *ap)
+	char		akey_enum_buf[ENUM_DESC_BUF] = {0};
+	char		akey[ENUM_KEY_BUF] = {0};
+	d_sg_list_t	sgl;
+	d_iov_t		iov;
+	daos_key_t	aiov;
+	daos_iod_t	iod;
+
+	/* loop to enumerate akeys */
+	while (!daos_anchor_is_eof(&akey_anchor)) {
+		memset(akey_kds, 0, sizeof(akey_kds));
+		memset(akey_enum_buf, 0, sizeof(akey_enum_buf));
+ 		memset(akey, 0, sizeof(akey));
+		akey_number = ENUM_DESC_NR;
+
+		sgl.sg_nr     = 1;
+		sgl.sg_nr_out = 0;
+		sgl.sg_iovs   = &iov;
+
+		d_iov_set(&iov, akey_enum_buf, ENUM_DESC_BUF);
+
+		/* get akeys */
+		rc = daos_obj_list_akey(*src_oh, DAOS_TX_NONE, &diov,
+					&akey_number, akey_kds,
+					&sgl, &akey_anchor, NULL);
+		if (rc != 0) {
+			fprintf(stderr, "failed to list akeys: %d\n", rc);
+			D_GOTO(out, rc);
+		}
+
+		/* if no akeys returned move on */
+		if (akey_number == 0)
+			continue;
+
+		/* parse out individual akeys based on key length and
+		 * numver of dkeys returned
+		 */
+		for (ptr = akey_enum_buf, j = 0; j < akey_number; j++) {
+			snprintf(akey, akey_kds[j].kd_key_len + 1, "%s", ptr);
+			if (buf_len >= ENUM_KEY_BUF) {
+				fprintf(stderr, "akey is too large: %d\n", rc);
+				D_GOTO(out, rc = -1);
+			}
+			d_iov_set(&aiov, (void *)akey, akey_kds[j].kd_key_len);
+
+			/* set iod values */
+			iod.iod_nr   = 1;
+			iod.iod_type = DAOS_IOD_SINGLE;
+			iod.iod_size = DAOS_REC_ANY;
+			iod.iod_recxs = NULL;
+
+			d_iov_set(&iod.iod_name, (void *)akey, strlen(akey));
+
+			/* do fetch with sgl == NULL to check if iod type
+			 * (ARRAY OR SINGLE VAL)
+			 */
+			rc = daos_obj_fetch(*src_oh, DAOS_TX_NONE, 0, &diov,
+					    1, &iod, NULL, NULL, NULL);
+			if (rc != 0) {
+				fprintf(stderr, "failed to fetch source "
+					"object: %d\n", rc);
+				D_GOTO(out, rc);
+			}
+
+			/* if iod_size == 0 then this is a DAOS_IOD_ARRAY
+			 * type
+			 */
+			if ((int)iod.iod_size == 0) {
+				rc = cont_copy_recx_array(&diov, &aiov, src_oh,
+						     dst_oh, &iod);
+				if (rc != 0) {
+					fprintf(stderr, "failed to copy "
+						"record: %d\n", rc);
+					D_GOTO(out, rc);
+				}
+			} else {
+				rc = cont_copy_recx_single(&diov, src_oh,
+						      dst_oh, &iod);
+				if (rc != 0) {
+					fprintf(stderr, "failed to copy "
+						"record: %d\n", rc);
+					D_GOTO(out, rc);
+				}
+			}
+			/* advance to next akey returned */
+			ptr += akey_kds[j].kd_key_len;
+		}
+	}
+out:
+	return rc;
+}
+
+static int
+cont_copy_list_dkeys(daos_handle_t *src_oh,
+		     daos_handle_t *dst_oh)
+{
+	int		rc = 0;
+	int		j = 0;
+	int		buf_len = 0;
+	char		*ptr;
+	daos_anchor_t	dkey_anchor = {0};
+	daos_key_desc_t	dkey_kds[ENUM_DESC_NR] = {0};
+	uint32_t	dkey_number = ENUM_DESC_NR;
+	char		dkey_enum_buf[ENUM_DESC_BUF] = {0};
+	char		dkey[ENUM_KEY_BUF] = {0};
+	d_sg_list_t	sgl;
+	d_iov_t		iov;
+	daos_key_t	diov;
+
+	/* loop to enumerate dkeys */
+	while (!daos_anchor_is_eof(&dkey_anchor)) {
+		memset(dkey_kds, 0, sizeof(dkey_kds));
+		memset(dkey_enum_buf, 0, sizeof(dkey_enum_buf));
+ 		memset(dkey, 0, sizeof(dkey));
```

Comment:
(style)  code indent should use tabs where possible
(style)  please, no space before tabs